tests/test_openeo_gfmap/test_unit_patch_extractors.py

```diff
@@ -0,0 +1,91 @@
+from openeo_gfmap.features import PatchFeatureExtractor
+from openeo.udf import XarrayDataCube
+
+import pytest
+import numpy as np
+import xarray as xr
+from unittest.mock import MagicMock, patch
+
+LAT_HARMONIZED_NAME = "GEO-LAT"
+LON_HARMONIZED_NAME = "GEO-LON"
+EPSG_HARMONIZED_NAME = "GEO-EPSG"
+
+class DummyPatchFeatureExtractor(PatchFeatureExtractor):
+    def output_labels(self):
+        return ["label1", "label2"]
+
+    def execute(self, inarr: xr.DataArray) -> xr.DataArray:
+        return inarr  # Simplified for testing purposes
+
+@pytest.fixture
+def mock_feature_extractor():
+    return DummyPatchFeatureExtractor()
+
+@pytest.fixture
+def mock_data_array():
+    return xr.DataArray(np.random.rand(10, 10), dims=["y", "x"])
+
+def test_get_latlons_epsg_none(mock_feature_extractor, mock_data_array):
+    mock_feature_extractor._epsg = None
+    with pytest.raises(Exception):
+        mock_feature_extractor.get_latlons(mock_data_array)
+
+def test_get_latlons_epsg_4326(mock_feature_extractor, mock_data_array):
+    mock_feature_extractor._epsg = 4326
+    result = mock_feature_extractor.get_latlons(mock_data_array)
+    assert LAT_HARMONIZED_NAME in result.coords['bands'].values
+    assert LON_HARMONIZED_NAME in result.coords['bands'].values
+
+@patch('pyproj.Transformer.from_crs')
+def test_get_latlons_reproject(mock_from_crs, mock_feature_extractor, mock_data_array):
```

Comment:
Where is the mock_from_crs defined? Or is that covered by the patch annotation? For test readability I think (should check) it is best practice to put the arguments in fixtures (either in the file itself or in a conftest is you need it in multiple files) or in a parametrize annotation.
Is there a reason you are mocking the pyproj package? I think it is important to test external packages, so if they push a breaking update we catch that in the tests early on.

Reply:
I'll clarify the mock from crs.

In terms of mocking, I disagree. When it comes to writing unit tests, my primary goal here is to test the openeo-gfmap code in isolation. I want to ensure that any failures in our tests are due to issues in our own code, not due to problems or changes in the external libraries we use.

Reply:
I like the idea of using mocks to isolate (large) aspects of our own code to ease debugging, but not when it comes to using external parties. Part of the importance of unit tests is also to test the integration and usage of external libs, as a change to their libs might break our own pipeline and e.g. deprecated warnings can also be noticed there.

Merely checking if a call is made might for example also not catch improper usage of (in this case) the pyproj package.